cabal-install/Distribution/Client/ProjectConfig.hs

```diff
@@ -879,7 +888,44 @@ readSourcePackage verbosity (ProjectPackageLocalDirectory dir cabalFile) = do
       packageSource        = LocalUnpackedPackage (root </> dir),
       packageDescrOverride = Nothing
     }
-readSourcePackage _verbosity _ =
+
+readSourcePackage verbosity distDirLayout (ProjectPackageRemoteRepo repo) = do
+    let
+      scmSrcDir  = "scm-" ++ showHashValue (hashSourceRepo repo)
+      destDir = distUnpackedSrcRootDirectory distDirLayout </> scmSrcDir
+
+      subdir  = fromMaybe "." (repoSubdir repo)
+      pkgdir  = destDir </> subdir
+
+      -- we only need this for error messages
+      repoloc = fromMaybe "" (repoLocation repo)
+
+    repoExists <- liftIO $ doesDirectoryExist destDir
+    when (not repoExists) $ do
+      branchers <- liftIO findUsableBranchers
+      case findBranchCmd branchers [repo] Nothing of
+        Just (BranchCmd fork) -> do
+          exitCode <- liftIO $ fork verbosity destDir
+          case exitCode of
+            ExitSuccess   -> return ()
+            ExitFailure _ -> fail ("Couldn't fork package from " ++ repoloc)
+        Nothing -> fail ("No usable brancher found for " ++ repoloc)
+
+    pkgdirExists <- liftIO $ doesDirectoryExist pkgdir
+    when (not pkgdirExists) $ do
+      fail ("Subdirectory does not exists in repository " ++ repoloc)
+
+    matches <- matchFileGlob (globStarDotCabal pkgdir)
+    case matches of
+        [cabalFile]
+            -> readSourcePackage
+                verbosity
+                distDirLayout
+                (ProjectPackageLocalDirectory pkgdir cabalFile)
```

Comment:
I don't think we want to convert this into a project local package, we want to preserve the information that this package is from a source repo and not necessarily treat it as local. That is we should be able to treat it like a package from hackage.

If we do this then it means we can put packages built from repos into the store and share them between multiple projects. This would be a big win. But it also means we need a mechanism to obtain a hash of the sources in the repo. This does not have to be a content hash of the unpacked dir, any stable content hash will do. So for git for example, the current git checkout hash is perfect. For other repo types other methods may be needed.